File: tests/Blokemon.Game.Tests/DeferredChoiceTests.cs
using Blokemon.Game;
using Shouldly;

namespace Blokemon.Game.Tests;

public sealed class DeferredChoiceTests
{
    [Test]
    public async Task OpponentChosenDiscard_PersistsUntilTheOpponentChoosesTheirHandCards()
    {
        var engine = MatchScenario.Engine();
        var initial = MatchScenario.BattleState(
            "BLK-024",
            "BLK-150",
            ["VIM-DODGY", "VIM-DODGY", "VIM-DODGY"],
            79
        );
        var first = MatchScenario.Card(
            "other-mitt-1",
            "BLK-004",
            MatchScenario.SecondPlayer,
            CardZone.Mitt,
            -1
        );
        var second = MatchScenario.Card(
            "other-mitt-2",
            "VIM-SOBER",
            MatchScenario.SecondPlayer,
            CardZone.Mitt,
            -1
        );
        initial = initial with
        {
            Cards = FrozenList<CardState>.Create(
                initial.Cards.Append(first).Append(second).OrderBy(static card => card.Id)
            ),
        };

        var requested = (CommandOutcome.Applied)
            engine.Apply(initial, MatchScenario.AttackCommand(initial, "BLK-024-B02"));
        var requirement = requested.State.PendingEffect!.Requirements.Single();
        var cpu = new DeterministicCpu();
        var decision = (CpuDecision.Selected)
            cpu.Choose(engine, requested.State, MatchScenario.SecondPlayer);
        var resolved = MatchScenario.Applied(
            engine.Apply(requested.State, decision.Action.Command)
        );

        requested.State.Phase.ShouldBe(MatchPhase.AwaitingEffectChoice);
        requirement.Chooser.ShouldBe(MatchScenario.SecondPlayer);
        requirement.Minimum.ShouldBe(2);
        resolved.Card(first.Id).Zone.ShouldBe(CardZone.EmptiesTray);
        resolved.Card(second.Id).Zone.ShouldBe(CardZone.EmptiesTray);
    }

    [Test]
    public async Task OpponentChosenSwitch_PersistsUntilTheOpponentChooses()
    {
        var engine = MatchScenario.Engine();
        var initial = MatchScenario.BattleState("BLK-012", "BLK-001", ["VIM-BLAZED"], 81);
        var bench = MatchScenario.Card(
            "defender-bench",
            "BLK-004",
            MatchScenario.SecondPlayer,
            CardZone.Booth,
            -1
        );
        initial = initial with
        {
            Cards = FrozenList<CardState>.Create(
                initial.Cards.Append(bench).OrderBy(card => card.Id)
            ),
        };

        var requested = (CommandOutcome.Applied)
            engine.Apply(initial, MatchScenario.AttackCommand(initial, "BLK-012-B01"));

        requested.State.Phase.ShouldBe(MatchPhase.AwaitingEffectChoice);
        requested.State.PendingEffect.ShouldNotBeNull();
        requested.State.Card(new CardInstanceId("defender")).Zone.ShouldBe(CardZone.Oche);

        var requirement = requested.State.PendingEffect!.Requirements.Single();
        var wrongChooser = new MatchCommand.ResolveEffectChoice(
            new CommandId("wrong-chooser"),
            requested.State.Id,
            MatchScenario.FirstPlayer,
            requested.State.Revision,
            FrozenList<EffectChoice>.Create(
                new EffectChoice.Cards(
                    requirement.Id,
                    FrozenList<CardInstanceId>.Create(new CardInstanceId("defender-bench"))
                )
            )
        );
        var rejected = (CommandOutcome.Rejected)engine.Apply(requested.State, wrongChooser);

        rejected.Rejection.Code.ShouldBe(CommandRejectionCode.WrongChooser);
        rejected.State.ShouldBe(requested.State);

        var cpu = new DeterministicCpu();
        var decision = (CpuDecision.Selected)
            cpu.Choose(engine, requested.State, MatchScenario.SecondPlayer);
        var resolved = (CommandOutcome.Applied)
            engine.Apply(requested.State, decision.Action.Command);

        decision.Action.Kind.ShouldBe(LegalActionKind.ResolveEffectChoice);
        resolved.State.PendingEffect.ShouldBeNull();
        resolved.State.Card(new CardInstanceId("defender-bench")).Zone.ShouldBe(CardZone.Oche);
        resolved.State.Card(new CardInstanceId("defender")).Zone.ShouldBe(CardZone.Booth);
    }
}

File: tests/Blokemon.Game.Tests/DeferredChoiceTests.fs
namespace Blokemon.Game.Tests

open System.Collections.Immutable
open Blokemon.Game
open FsUnit
open TUnit.Core

[<AutoOpen>]
module private DeferredChoiceFixtures =

    let selected (decision: CpuDecision) =
        match decision with
        | CpuDecision.Selected action -> action
        | CpuDecision.NoLegalAction -> failwith "Expected a legal action to be available."

type DeferredChoiceTests() =

    [<Test>]
    member _.``an opponent-chosen switch should wait until the opponent chooses and should refuse the wrong chooser``
        ()
        =
        let engine = MatchScenario.Engine()
        let initial = MatchScenario.BattleState "BLK-012" "BLK-001" [ "VIM-DODGY" ] 81UL

        let bench =
            MatchScenario.PlainCard
                "defender-bench"
                "BLK-004"
                MatchScenario.SecondPlayer
                CardZone.Booth
                -1

        let initial = MatchScenario.WithCards initial [ bench ]

        let requested =
            MatchScenario.Applied(
                engine.Apply(initial, MatchScenario.AttackCommand initial "BLK-012-B01")
            )

        requested.Phase |> should equal MatchPhase.AwaitingEffectChoice
        requested.PendingEffect.IsSome |> should be True
        (requested.Card(CardInstanceId "defender")).Zone |> should equal CardZone.Oche

        let requirement = requested.PendingEffect.Value.Requirements |> Seq.exactlyOne

        let wrongChooser =
            MatchScenario.Command
                requested
                "wrong-chooser"
                MatchScenario.FirstPlayer
                (ImmutableArray.Create(
                    EffectChoice.Cards(
                        requirement.Id,
                        ImmutableArray.Create(CardInstanceId "defender-bench")
                    )
                ))
                MatchAction.ResolveEffectChoice

        let rejectedState, rejection =
            MatchScenario.Rejected(engine.Apply(requested, wrongChooser))

        rejection.Code |> should equal CommandRejectionCode.WrongChooser
        rejectedState |> should equal requested

        let cpu = DeterministicCpu()
        let decision = selected (cpu.Choose(engine, requested, MatchScenario.SecondPlayer))
        let resolved = MatchScenario.Applied(engine.Apply(requested, decision.Command))

        decision.Kind |> should equal LegalActionKind.ResolveEffectChoice
        resolved.PendingEffect.IsNone |> should be True

        (resolved.Card(CardInstanceId "defender-bench")).Zone
        |> should equal CardZone.Oche

        (resolved.Card(CardInstanceId "defender")).Zone |> should equal CardZone.Booth
